crates/ruff_linter/src/rules/ruff/rules/none_not_at_end_of_union.rs

```diff
@@ -0,0 +1,74 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::Expr;
+use ruff_python_semantic::analyze::typing::traverse_union;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for type annotations where `None` is not at the end of an union.
+///
+/// ## Why is this bad?
+/// Type annotation unions are associative, meaning that the order of the elements
+/// does not matter. The `None` literal represents the absence of a value. For
+/// readability, it's preferred to write the more informative type expressions first.
+///
+/// ## Example
+/// ```python
+/// def func(arg: None | int): ...
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def func(arg: int | None): ...
+/// ```
+///
+/// ## References
+/// - [Python documentation: Union type](https://docs.python.org/3/library/stdtypes.html#types-union)
+/// - [Python documentation: `typing.Optional`](https://docs.python.org/3/library/typing.html#typing.Optional)
+/// - [Python documentation: `None`](https://docs.python.org/3/library/constants.html#None)
+#[violation]
+pub struct NoneNotAtEndOfUnion;
+
+impl Violation for NoneNotAtEndOfUnion {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "`None` not at the end of the type annotation.".to_string()
+    }
+}
+
+/// RUF036
+pub(crate) fn none_not_at_end_of_union<'a>(checker: &mut Checker, union: &'a Expr) {
+    let semantic = checker.semantic();
+    let mut none_exprs: Vec<&Expr> = Vec::new();
```

Comment:
Nit: Consider using a smallvec here so that we avoid allocating in all valid cases (where there is a single `None`) at the end.